storage/provider/openstack.go

```diff
@@ -0,0 +1,399 @@
+func (p *OpenstackProvider) VolumeSource(environConfig *config.Config, providerConfig *storage.Config) (storage.VolumeSource, error) {
+
+	if err := p.ValidateConfig(providerConfig); err != nil {
+		return nil, err
+	}
+
+	openstackClient, err := p.ClientFactory(providerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &openstackVolumeSource{openstackClient}, nil
+}
+
+// FilesystemSource implements storage.Provider.
+func (p *OpenstackProvider) FilesystemSource(environConfig *config.Config, providerConfig *storage.Config) (storage.FilesystemSource, error) {
+	return nil, errors.NotSupportedf("filesystems")
+}
+
+// Supports implements storage.Provider.
+func (p *OpenstackProvider) Supports(kind storage.StorageKind) bool {
+	switch kind {
+	case storage.StorageKindBlock:
+		return true
+	}
+	return false
+}
+
+// Scope implements storage.Provider.
+func (s *OpenstackProvider) Scope() storage.Scope {
+	panic("not implemented")
+}
+
+// ValidateConfig implements storage.Provider.
+func (p *OpenstackProvider) ValidateConfig(cfg *storage.Config) error {
+
+	for _, cfgKey := range []string{
+		openstackCfgKeyUrl,
+		openstackCfgKeyUser,
+		openstackCfgKeySecrets,
+		openstackCfgKeyRegion,
+		openstackCfgKeyTenantName} {
+		if _, ok := cfg.Attrs()[cfgKey]; !ok {
+			return errors.NotAssignedf("requisite configuration was not set: %v", cfgKey)
+		}
+	}
+
+	return nil
+}
+
+// Dynamic implements storage.Provider.
+func (p *OpenstackProvider) Dynamic() bool {
+	panic("not implemented")
+}
+
+type openstackVolumeSource struct {
+	openstackClient OpenstackAdapter
+}
+
+var _ storage.VolumeSource = (*openstackVolumeSource)(nil)
+
+// CreateVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) CreateVolumes(args []storage.VolumeParams) ([]storage.Volume, []storage.VolumeAttachment, error) {
+
+	vols := make([]storage.Volume, 0, len(args))
+	volAtt := make([]storage.VolumeAttachment, 0, len(args))
+
+	for _, a := range args {
+		if a.Provider != CinderProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				a.Provider,
+			)
+			continue
+		}
+
+		createdVol, err := s.openstackClient.CreateVolume(a.Size, a.Tag.String())
+		if err != nil {
+			return nil, nil, err
+		}
+		vols = append(vols, createdVol)
+
+		if a.Attachment != nil {
+
+			a.Attachment.VolumeId = createdVol.VolumeId
+
+			attachments, err := s.AttachVolumes([]storage.VolumeAttachmentParams{*a.Attachment})
+			if err != nil {
+				return nil, nil, err
+			}
+
+			volAtt = append(volAtt, attachments...)
+		}
+	}
+
+	return vols, volAtt, nil
+}
+
+// DescribeVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DescribeVolumes(volIds []string) ([]storage.Volume, error) {
+
+	// In most cases, it is quicker to get all volumes and loop
+	// locally than to make several round-trips to the provider.
+	cinderVols, err := s.openstackClient.GetVolumesSimple()
```

Comment:
I don't think there is. You can get detailed volume information via /v2/​{tenant_id}​/volumes/​{volume_id}​, but I figured 1 batch request would be faster than several API calls.